### ISSUE TITLE
Pex stripping querystring from artifact downloads

### ISSUE THREAD

**Issue by jake-sigtech** [pantsbuild/pex#2156]
Pex strips querystring from artifact URL [code](https://github.com/pantsbuild/pex/blob/adf7e5449562fc819a3e93c0f4f4d6e155284af0/pex/resolve/resolved_requirement.py#L84-L86)

We are using [pypicloud](https://pypicloud.readthedocs.io/en/latest/) for a private python package index.
This is backed by S3 and returns package links as s3 pre-signed urls, which contain querystrings for authentication with s3.

Hence the artifact download fails, when the querystrings are stripped (403 error).

Would it be possible not to strip the querystring?
The querystring changes each time, so would be best for caching if the url without the querystring is the one used for the cache.

**Comment by jsirois** [comment]
@jake-sigtech re:
> The querystring changes each time, so would be best for caching if the url without the querystring is the one used for the cache.

What exactly do you mean by "The querystring changes each time,"? Each second, minute, after each download, after each GET or HEAD ... the lock file only has 1 place to store the URL of an artifact for all time; so if the URLs are not stable, the lock will just naturally become invalid and that is not good.

**Comment by jsirois** [comment]
@jake-sigtech thanks but that doesn't really answer the key question, what about that expires in the URL, is it sufficiently far in the future to write it down in a lockfile today and have it work next year?

**Comment by cognifloyd** [comment]
That means you cannot effectively use a lockfile. To use a lockfile, you must have a permanent url available somehow.

If we did preserve the query string, then your lockfile would only be valid for the duration of the pre-signed url(s) which would render it effectively useless.

Off-topic aside: Are you using pulp project to host your internal pypi or something else?

**Comment by thejcannon** [comment]
You will find only pain using AWS CodeArtifact if you're using Pants.
https://pantsbuild.slack.com/archives/C046T6T9U/p1687276300085429

We ended up circumventing the pain by not using CodeArtifact (and just using S3) and forking Pex and adding a sliver of support to have it re-write S3 URLs with auth query params before sending them off to `pip` for download.

If you're only using Pex, there are a few ways to get CodeArtifact working, however the _maximum_ token timeout of 12 hours is the crux of all pain.

**Comment by thejcannon** [comment]
I guess we're talking about 2 different things now:

1. How URLs are handled by the tools. `pip` basically requires that URLs it is given are accessible through just the URL alone. If you need auth, it MUST come from HTTP basic auth (`username:password@<website>`). So, by wrapping `pip`, Pex also has this behavior. Pex does have some of its own logic, though. Namely that the lockfile can lock without basic auth (so that nothing is leaked) and then allows for auth loading via `.netrc`. After injecting auth, this is still a URL without headers so `pip` is happy to download.
FWIW @jsirois `pip` doesn't strip query params (as evidenced by local testing and my fork working) so is there a reason Pex does? If not, I can easily make a PR that fixes this issue.

2. Authentication. Services that serve packages and require auth are pigeonholed into the requirement by `pip` that a simple URL is sufficient (it's a reasonable thing to assume). So your choices are:
    -  Choosing a provider that supports basic auth, and using env vars or other tricks to have Pants pass the right `--find-links`. When the auth is expired and changed, Pants' process cache is invalidated, so not ideal.
    - Choosing an auth provider that supports basic auth, and using `.netrc` to store the auth. When the auth is expired and changed, Pants' cache _isn't_ invalidated. So more ideal.
    - Forking Pex. At this point anything is possible. I recommend doing as little changing as possible though. For us that means storing artifacts in S3 and quickly generating a pre-signed S3 URL to hand to `pip`. Because query params are how auth is done, the Pex code that hands a URL to `pip` is still valid.

**Comment by jsirois** [comment]
> FWIW @jsirois pip doesn't strip query params (as evidenced by local testing and my fork working) so is there a reason Pex does? If not, I can easily make a PR that fixes this issue.

@jake-sigtech and I discussed this in slack, but no, not really. Pex only cares that it strips the fragment since that carries (ied) fingerprint data in the form `#sha256=abcd1234`. As you clearly understand and I think Jake does now too, that does not help Jake's case here where the fundamental issue is impermanent URLs in a lock format that requires / assumes exactly permanent URLs.

> Forking Pex. ...

Please never do this unless either:
+ You know what you're doing is crazy and could never be a part of the project you're forking (extreme ugly pragmatics).
+ The project is intransigent to your reasonable proposal / request / etc.

So clearly Pex could grow a more rich auth system to re-write URLs more generally beyond the .netrc support it currently has. Something along the lines of supplying `--auth <URL PREFIX>=qp=<QUERY PARAMS> --auth <URL PREFIX>=header=<HEADER>`.

**Comment by thejcannon** [comment]
Yeah, I'm not saying I have an actual use-case, just saying that there could be one and then this _would_ be a bug. If the intent is to have Pex match pip, this is a bug, use-case nonwithstanding. And one with a very simple fix.

**Comment by thejcannon** [comment]
And yeah I wouldn't suggest Pex do S3 auth handling, which is why I didn't bother with an issue/PR.

S3 auth works where the auth is time-based and specific to the URL. That means you can't pass in a pre-defined value. The "signed" URL must be generated at the time of the request. Additionally, knowing how to "sign" the request is best left to AWS libraries (botocore) so you'd be embedding it and all its deps into PEX for this narrow use case which isn't supported by pip either.

The patch itself is narrow once you stomach those nasty changes, so I'm really not worried about maintenance.

**Comment by jsirois** [comment]
You've got all that right @jake-sigtech. Pex does parse the URL for the wheel file name as well:
https://github.com/pantsbuild/pex/blob/0e4817469360673c4516637a743bcda54922a94c/pex/resolve/locker.py#L296-L310

@jake-sigtech I think you're missing the fundamental issue. As designed, Pex locking currently _requires_ a stable for all of time URL to use to download artifacts when using a lock file. The URLs your system uses are not stable for all time. That is the exact issue here.

To elaborate - your system requires URL parameters for authentication and those URL parameters _change_ with time; thus your URLs are not stable for all time and cannot work with Pex locking as it stands today. In order to make this sort of system work, Pex would need to grow extra authentication logic that would know how to inject the unstable portion of the URL just-in-time when downloading artifacts from the lock file (where just the stable portion of the URL would be stored).

**Comment by jsirois** [comment]
> Could you explain why a stable URL for all time is important?

It is currently important for speed of downloads.

With a fixed URL in the lock file resolved at lock creation time, lock use just involves selecting which artifacts to use from the lock file @ O(~100ms), and then downloading that set of artifacts in parallel. Currently (since Pex 2 was cut), Pex relies on Pip to do all authentication / resolving / downloading / etc. Pex 1 had a bespoke resolver that was a large source of bugs; so Pex un-learning things to do with resolving artifacts was a goal. Pip is slow when asked to resolve a dependency via a requirement string even when passed `--no-deps`, but it's fast to download a specific URL.

> My (limited) understanding from locking mechanisms in other languages is that the key part is the file hash.
>
> In theory I could imagine a solution where no URL is provided and pip/pex have to search in a package index for a file that matches that hash?

Yes, Pex could do that either at the cost of speed explained above or at the cost of learning how to interact with indexes (PEP-503) or both.

Hopefully it is clear that Pex cannot currently handle your index with credentials that are both part of the artifact URL _and_ vary over time. It is absolutely possible to handle this, but it requires re-work of Pex locking as mentioned previously to learn more about one or more things Pip currently handles for it.

**Comment by thejcannon** [comment]
Regardless of what the query string is used for, is there any reason PEX shouldn't support it if pip does?

**Comment by jsirois** [comment]
@thejcannon no (as mentioned here: https://github.com/pantsbuild/pex/issues/2156#issuecomment-1615397487). So fixing the latent bug is fine but it absolutely does not help @jake-sigtech or anyone else that I'm aware of.

**Comment by thejcannon** [comment]
Yup, just wanted make sure there wasn't something nonobvious I was mising

**Comment by jsirois** [comment]
Alright, this still doesn't help @jake-sigtech IIUC but should help some situations - I'm picking up #336 / #838 to allow specifying `--use-pip-config` in Pex commands to disable the current unconditional use of Pip `--isolated` mode. In the immediate it will help Pex tests switch from PyPI to a local `devpi-server` proxy / cache which saves > 50% of IT runtime in my current testing.

**Comment by jsirois** [comment]
This is getting knocked out as part of fixing #2348. It's not needed for that fix, but the logic of splitting a raw URL from a download URL (new) and a normalized keyable URL is in that PR and makes this fix well-contained.
